src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -3516,26 +3522,56 @@ private TypeSymbol GetStackAllocType(SyntaxNode node, TypeWithAnnotations elemen
         {
             var inLegalPosition = ReportBadStackAllocPosition(node, diagnostics);
             hasErrors = !inLegalPosition;
-            if (inLegalPosition && !node.IsLocalVariableDeclarationInitializationForPointerStackalloc())
+            if (inLegalPosition && !isStackallocTargetTyped(node))
             {
                 CheckFeatureAvailability(node, MessageID.IDS_FeatureRefStructs, diagnostics);
 
                 var spanType = GetWellKnownType(WellKnownType.System_Span_T, diagnostics, node);
-                if (!spanType.IsErrorType())
-                {
-                    return ConstructNamedType(
-                        type: spanType,
-                        typeSyntax: node.Kind() == SyntaxKind.StackAllocArrayCreationExpression
-                            ? ((StackAllocArrayCreationExpressionSyntax)node).Type
-                            : node,
-                        typeArgumentsSyntax: default,
-                        typeArguments: ImmutableArray.Create(elementTypeWithAnnotations),
-                        basesBeingResolved: null,
-                        diagnostics: diagnostics);
-                }
+                return ConstructNamedType(
+                    type: spanType,
+                    typeSyntax: node.Kind() == SyntaxKind.StackAllocArrayCreationExpression
+                        ? ((StackAllocArrayCreationExpressionSyntax)node).Type
+                        : node,
+                    typeArgumentsSyntax: default,
+                    typeArguments: ImmutableArray.Create(elementTypeWithAnnotations),
+                    basesBeingResolved: null,
+                    diagnostics: diagnostics);
             }
 
+            // We treat the stackalloc as target-typed, so we give it a null type for now.
             return null;
+
+            // Is this a context in which a stackalloc expression could be converted to the corresponding pointer
+            // type? The only context that permits it is the initialization of a local variable declaration (when
+            // the declaration appears as a statement or as the first part of a for loop).
+            static bool isStackallocTargetTyped(SyntaxNode node)
+            {
+                Debug.Assert(node != null);
+
+                SyntaxNode equalsValueClause = node.Parent;
+
+                if (!equalsValueClause.IsKind(SyntaxKind.EqualsValueClause))
```

Comment:
> if (!equalsValueClause.IsKind(SyntaxKind.EqualsValueClause)) [](start = 15, length = 61)

Could it be parenthesized?

Reply:
No

Reply:
> No 

Do we have a test for a scenario like that?

---
In reply to: [359103173](https://github.com/dotnet/roslyn/pull/40457#discussion_r359103173) [](ancestors = 359103173)

Reply:
Yes.  Customers depend on it.  See for example Nikov's latest comment on Twitter yesterday.

---
In reply to: [359106011](https://github.com/dotnet/roslyn/pull/40457#discussion_r359106011) [](ancestors = 359106011,359103173)

Reply:
> Yes.  Customers depend on it.  See for example Nikov's latest comment on Twitter yesterday.

Could you please point to the existing unit-test?

---
In reply to: [359107068](https://github.com/dotnet/roslyn/pull/40457#discussion_r359107068) [](ancestors = 359107068,359106011,359103173)

Reply:
`StackAllocInitializerTests.StackAllocInParenthesis`

---
In reply to: [359107823](https://github.com/dotnet/roslyn/pull/40457#discussion_r359107823) [](ancestors = 359107823,359107068,359106011,359103173)